client/app/react-components/DynamicTable.js

```diff
@@ -0,0 +1,347 @@
+    preparedRows: [],
+    rowsToDisplay: [],
+    totalPages: 1,
+    currentPage: 1,
+    searchTerm: '',
+    orderBy: [],
+    // eslint-disable-next-line react/no-unused-state
+    rows: [],
+    columns: [],
+  }
+
+  static getDerivedStateFromProps(newProps, oldState) {
+    let shouldSortRows = false;
+    const newState = { ...oldState };
+
+    if (newProps.columns !== oldState.columns) {
+      Object.assign(newState, {
+        columns: newProps.columns,
+        orderBy: [],
+        currentPage: 1,
+        searchTerm: '',
+        searchColumns: filter(newProps.columns, 'allowSearch'),
+      });
+      shouldSortRows = true;
+    }
+    if (newProps.rows !== oldState.rows) {
+      newState.rows = newProps.rows;
+      newState.currentPage = 1;
+      newState.totalPages = calculateTotalPages(newProps.rows.length, newProps.itemsPerPage);
+      newState.pages = getPages(1, newState.totalPages, 6);
+      shouldSortRows = true;
+    }
+
+    if (newProps.itemsPerPage !== oldState.itemsPerPage) {
+      newState.itemsPerPage = validateItemsPerPage(newProps.itemsPerPage);
+      newState.currentPage = 1;
+    }
+    if (shouldSortRows) {
+      newState.preparedRows = sortRows(
+        filterRows(
+          newState.rows,
+          oldState.searchTerm,
+          oldState.searchColumns,
+        ),
+        newState.orderBy,
+      );
+    }
+    newState.rowsToDisplay = displayRows(newState.preparedRows, newState.currentPage, newState.itemsPerPage);
+    return newState;
+  }
+
+  selectPage = (pageNumber, event) => {
+    if (event) {
+      event.preventDefault();
+    }
+    if (pageNumber !== this.state.currentPage && pageNumber > 0 && pageNumber <= this.state.totalPages) {
+      if (event && event.target) {
+        event.target.blur();
+      }
+      this.setState({
+        currentPage: pageNumber,
+        pages: getPages(pageNumber, this.state.totalPages, 6),
+        rowsToDisplay: displayRows(this.state.preparedRows, pageNumber, this.state.itemsPerPage),
+      });
+    }
+  }
+
+  render() {
+    const noNext = this.state.currentPage === this.state.totalPages;
+    const noPrev = this.state.currentPage === 1;
+    const goNext = e => this.selectPage(this.state.currentPage + 1, e);
+    const goPrev = e => this.selectPage(this.state.currentPage - 1, e);
+    return (
+      <React.Fragment>
+        <div className="dynamic-table-container">
+          <table className="table table-condensed table-hover">
+            <thead>
+              <tr>
+                {this.props.columns.map(c => (
+                  <ColumnHeader
+                    key={c.name}
+                    column={c}
+                    orderBy={this.state.orderBy}
+                    onClick={this.onColumnHeaderClick}
+                  />
+                ))}
+                <th className="dynamic-table-spacer" />
+              </tr>
+            </thead>
+            {this.state.searchColumns.length > 0 ?
+              <thead>
```

Comment:
It's only valid to have one `thead` per table. You could have as many `tr`s within the `thead` as you want, though.

Reply:
this was ported from [dynamic-table](https://github.com/mozilla/redash/blob/master/client/app/components/dynamic-table/dynamic-table.html#L3-L26).
![it was like this when I found it](https://www.pleated-jeans.com/wp-content/uploads/2012/03/2f0b476e9f0e50be2afe3c33f248b3d9059ab1cc.gif)